Merge the `sn-node-manager` crate into the `safe_network` workspace

jacderida opened the pull request:
This retains the commit history from the node manager repository.

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 23 Jan 24 20:20 UTC
This pull request includes multiple changes across different files. 

1. The README.md file has undergone minor modifications to improve the clarity of instructions and content.
2. The Justfile script has been enhanced to include new targets for creating and deleting droplets, as well as modifying an existing target for error handling.
3. The CHANGELOG.md file documents additions, fixes, and enhancements to the project.
4. The Cargo.toml file in the sn-node-manager directory specifies package details, dependencies, and features.
5. A new script, run_local_service_network.sh, has been added to start a local service network and perform various operations.
6. A new file, helpers.rs, has been added containing helper functions for downloading and extracting release binaries.
7. The Vagrantfile has been added to configure a Vagrant virtual machine for the project.
8. The e2e.rs file includes a new test for installing and controlling services.
9. The Cargo.toml file includes a new "sn-node-manager" module in the members list.
10. The README.md file in the sn-node-manager directory provides information about Safenode Manager.
11. A new file, config.rs, has been added to handle file paths and permissions in Unix and Windows environments.
12. Some environment variables and CI job changes have been made in one of the files.
13. Some files and directories have been added to the ignored files list.

These changes span across various aspects such as documentation, scripts, configuration, tests, and code enhancements. Let me know if you need more details or assistance with any specific part.
<!-- reviewpad:summarize:end -->

davidrusu commented on sn_api/src/app/keys.rs:
```diff
@@ -102,9 +102,7 @@ mod tests {
-                return Ok(());
-            }
+            Keypair::Bls(_) => Ok(()),
             _ => {
                 return Err(eyre!("A BLS keypair should be generated by default."));
             }
```
Apply the same change to the err arm as well?
```suggestion
            _ => Err(eyre!("A BLS keypair should be generated by default."))
```

jacderida replied:
Ah, thanks. Clippy only flagged the `Ok` one. Will change this.

davidrusu reviewed (approved):
Looks good :ship:

davidrusu commented:
bors r+

jacderida committed:
chore: node manager initial commit

jacderida committed:
feat: provide `install` command

The install process is:

* Get the latest version number for `safenode`.
* Download a `safenode` archive for the current platform, using that version number.
* Extract it to a system wide location.
* Create a user for running the service if it doesn't already exist.
* Obtain a free port for the node service.
* Obtain a free port for the RPC service.
* Create a service for `safenode`, which will run the node and the RPC service.

This does *not* start any of the installed services--that will be a separate command.

The testing approach here is behaviour-driven unit tests, rather than integration tests, though I do
plan to add an integration test which will function as more of an end to end test that will grow
over time. The unit tests can be more numerous and cover small differences in configuration without
a full setup, which will require root access. The testing approach influenced the design here, with
the `install` function accepting traits which can be supplied mocks during the tests.

This commit also defines a merge workflow which runs the standard build and unit test jobs.

jacderida committed:
test: provide initial integration tests

What's provided in these tests is a first step, but the intention is for them to evolve into end to
end tests that will install, start/stop services, add new services, and so on. They will run by
executing the safenode-manager binary.

These tests are intended to demonstrate that we are running on our supported platforms without
introducing any major regressions. They will follow one config path without delving into different
configuration options--we'll leave that to the unit tests.

They need to run as the root user, and ideally, in a CI environment with a fresh build agent,
because they will create real services and not attempt to clean themselves up. Run on your own
machine at your own risk.

jacderida committed:
feat: provide `start` command

Start installed safenode services. If a service name or peer ID is not supplied, it will attempt to
start everything.

If the service status is `Installed` (meaning it has not been started for the first time since it
was installed) or `Stopped`, there will be an attempt to start it. If the service is `Running` we
will check again that it is actually running, and if it isn't, attempt to start it again.

Also change the integration tests here to one test which will be cross platform service
manipulation. The commands should work on different operating systems. Once a `status` command has
been added we will be able to use the application to validate itself.

jacderida committed:
chore: specify root and log dirs at install time

To facilitate starting a node again with the same data, we need to supply the `--root-dir` argument
at install time. In addition to that, we need to actually create the directory and assign it the
correct permissions such that the user running the service will be able to write to it.

The log dir is also specified. This makes the logs available at system-wide locations rather than
the default directories, which is the data directory of the user running the service, which can make
them difficult to access, especially on Windows. Despite being at a system-wide location, on Linux,
the directory is assigned permissions to be written to by the `safe` user.

I tested the configurations on Windows and Linux.

I have also added a utility Vagrantfile here for quick testing. The reason for using a VM is because
you need to run commands as root and actually install services. I didn't want to have to continually
clean up that mess on my dev machine.

jacderida committed:
chore: reference `sn_node_rpc_client` crate

Instead of introducing a new `sn_rpc_client` crate, we decided to make the existing
`sn_node_rpc_client` crate both a lib and a binary crate.

An update to the service manager crate was also merged.

jacderida committed:
feat: provide `stop` command

A fairly simple command that will stop a running service.

Sometimes you can encounter difficulty when trying to stop a service gracefully, but I decided to
only attempt to stop the service once and leave any retries to the service infrastructure.

jacderida committed:
feat: provide `status` command

The command provides both a summary and a detailed view of installed services.

It checks to see whether the service process is still running, because it's possible the process
could have died since the last time we checked.

The data and log directories were added to the `InstalledNode` struct so that we could keep track of
these and display them in the detailed status view since it seems like useful information to have.

jacderida committed:
test: extend the e2e test for new commands

With `stop` and `status` commands, we can have a more elaborate test that exercises all the
important functionality in the node manager.

Using the status reported by the application allows us to make the test cross platform.

jacderida committed:
ci: release process and licensing

Provide version bumping and release workflows. These are copied from the `safeup` repository; they
should serve fine for what we need here.

Also prepare the crate for publishing by adding a README and a license. The license verification
tool has also been enabled for CI. All source files have had copyright statement applied.

actions-user committed:
chore(release): 0.1.0

jacderida committed:
chore: improve docs for `start` and `stop` commands

Also add directories used in the deploy process to the .gitignore. Without these the publishing
process will fail.

actions-user committed:
chore(release): 0.1.1

jacderida committed:
refactor: rename `install` command to `add`

The word `add` more accurately reflects what is happening here, since the command can add any number
of services using the `--count` argument, and it is also intended to be used multiple times; you
don't really run an installation more than once.

jacderida committed:
feat: provide `--peer` argument

The `add` command is extended to provide a `--peer` argument. It uses the `sn_peers_acquisition`
crate and thus parses them in the exact same way as `safenode`.

jacderida committed:
feat: provide `remove` command

Removes an added node, which will delete its data and log directories. The node is marked as
removed rather than actually being removed from the node registry just because the service number is
incremented by one each time a new node is added, so it might be a little odd if numbers were
missing in the sequence.

actions-user committed:
chore(release): 0.1.2

actions-user committed:
chore(release): 0.1.3

actions-user committed:
chore(release): 0.1.4

jacderida committed:
feat: each service instance to use its own binary

The `add` command will download a copy of `safenode`, and if no version was specified, it will be
the latest version at the time the command was run. If the command is used a day later to add
another service, that could then be a different version. If both nodes were using the same binary at
the same location, one would inadvertently be upgraded. There may also be issues with updating a
binary that is running.

For the upgrade process, it also seems logical and safer for each service instance to have its own
binary. If the upgrade process failed for one service, for whatever reason, it wouldn't leave
anything else in an inconsistent state.

It's also possible at some point in the future that it could be desirable for people to run
different node versions. Perhaps they would want to test a new version before moving from something
stable.

This also fixes a bug with the `--peer-id` argument where it will return an empty list of peers if
there is an error. This allows for no peers to be specified, which is necessary in the case of the
genesis node.

jacderida committed:
feat: provide `upgrade` command

The upgrade process simple:
* Get the latest version
* Download the new binary
* Stop the running node
* Overwrite the old binary with the new binary
* Start the node

It avoids doing anything if all the nodes are already at the latest version, and it also only
downloads the new binary once, even if there are multiple nodes to upgrade.

Also provided here is a Bash script that can spin up a quick test machine on Digital Ocean. This is
useful for adding nodes that connect to real networks without having to be concerned about NAT on
the local VM. I tried to use the DO provider for Vagrant, but sadly it didn't work correctly; I
think it was out of date with the current API.

The branch for the `remove` command was also rebased on top of this commit. As a result, the
`safenode_path` on the `Node` struct was also set to an optional field to make it consistent with
`data_dir_path` and `log_dir_path`.

actions-user committed:
chore(release): 0.1.5

jacderida committed:
fix: accommodate service restarts in `status` cmd

If the service is configured to support restarts, the `safenode` process can die and then be started
again as a new process. In this case, the node in the registry needs to be reassigned with a new
PID. We can use the RPC service to obtain the new PID in a cross platform way.

jacderida committed:
feat: provide `--url` argument for `add` command

The URL enables use of a custom `safenode` binary that was built and uploaded somewhere. The URL
must point to a zip or gzipped tar archive which contains the binary.

The main reason for this is to support the creation of testnets that want to use a `safenode` binary
that is built from a fork of the main repository.

actions-user committed:
chore(release): 0.1.6

actions-user committed:
chore(release): 0.1.7

actions-user committed:
chore(release): 0.1.8

jacderida committed:
feat: custom port arguments for `add` command

When we're launching a genesis node, it needs to run on known ports, so we need to be able to
specify them.

It only really makes sense to use these arguments if a single service is being added, since multiple
services can't use the same port. So the code asserts that they can only be specified when one
service is being added.

actions-user committed:
chore(release): 0.1.9

jacderida committed:
feat: provide `run` command

NOTE: this commit also contains the corresponding `kill` command. I accidentally amended it and
couldn't untangle the changes from it without losing them.

Spin up `safenode` processes, along with a faucet, to run a local network. These processes are not
defined as services because they are not intended to be long running; they are for experimentation.

As in the `testnet` binary, the network is validated, but unlike in `testnet` the logs are not
inspected. We just query each RPC service to check that peers are connected.

At present you need to supply paths to both binaries, but this will be extended shortly to provide
the ability to run downloaded binaries.

The node registry was extended to keep track of the faucet process so it can easily be cleaned up in
a subsequent `kill` command.

The `status` command is also extended to accommodate the local network.

actions-user committed:
chore(release): 0.1.10

jacderida committed:
feat: download binaries in absence of paths

If `--node-path` or `--faucet-path` arguments are not supplied, the latest versions of released
binaries will be downloaded. This enables `safenode-manager run` without any arguments. The ability
to get specific versions is available through `--node-version` and `--faucet-version` arguments.

The `kill` command was also improved to accommodate a scenario where the local network doesn't spin
up properly; you can use it to clean up an incomplete network.

actions-user committed:
chore(release): 0.1.11

jacderida committed:
chore: use `--first` arg for genesis node

A new `--first` argument has been added to `safenode` to facilitate the distinction between when we
launch a genesis node and when we want the default network contacts to be used.

We now update the local `run` command to use this argument to correctly launch a genesis node.

Also updated the year in the copyright notices on all the source files. This is something that's not
legally necessary, but it's the way the licensing verification tool works.

jacderida committed:
chore: upgrade safe dependencies

Both `sn_node_rpc_client` and `sn_peers_acquisition` had MINOR version upgrades.

jacderida committed:
refactor: use clap for conflicting arguments

I only recently became aware of this feature, which allows you to specify conflicting arguments.
Using the feature removes some code I had that was doing this manually.

actions-user committed:
chore(release): 0.1.12

jacderida committed:
feat: provide `join` command

Add more nodes to an existing local network. It can be a network that's managed by the node manager,
or one that's been started independently. The latter scenario exists in one of our CI workflows for
the `safe_network` repo.

Whether you're joining a network managed by the node manager or not is determined by the use of the
`--peers` argument, which is why it's configured as an `Option`.

jacderida committed:
chore: respond to feedback

* Use clap to provide default node count
* Mark `join` arguments as conflicting
* Fix bug with single node on join
* Use `find_map` rather than `find` to eliminate additional step

jacderida committed:
fix: apply to correct argument

The conflict was applied to the wrong argument.

actions-user committed:
chore(release): 0.1.13

jacderida committed:
feat: apply `--first` argument to added service

If the `--first` argument is specified, it can only be when a single node is being added. The node
is then designated as the genesis node in the registry. The same also applies to the local setup.

jacderida committed:
chore: manually parse environment variable

On the `join` command, the peers argument is set as an option so we are not required to specify any
peers when adding additional nodes to an existing local network. When the argument is set as an
option, it doesn't parse the environment variable if it's set, so unfortunately we also have to
provide some manual parsing.

The commit also removes the node registry definition, which has been moved to `sn_protocol`. The
churn tests in `safe_network` CI now need to be able to parse the node manager's deployment
inventory so they can restart local nodes.

actions-user committed:
chore(release): 0.1.14

actions-user committed:
chore(release): 0.1.15

jacderida committed:
chore: tidy peer management for `join` command

The `peers` argument is no longer declared with an `Option` and the manual parsing of `SAFE_PEERS`
is removed. If the user doesn't specify any peers using `--peer` or `SAFE_PEERS`, the
`get_peers_from_args` function will return a particular type of error. In this case, we can just
assume the user wants the nodes to join an existing local network.

actions-user committed:
chore(release): 0.1.16

jacderida committed:
chore: additional info in `status` cmd

The binary path is provided because this is useful to know, particularly for nodes installed as
services.

The number of connected peers are also included, because again, this can be useful to know if the
network has initialised itself correctly, and how it might change over time.

There was a bit of a refactor in here to have the `run_node` function return a `Node` rather than
a multiaddr, because there's a function on the node now to get the multiaddr. The refactor also
eliminated a test case, which was no longer necessary.

joshuef committed:
feat: add quic/tcp features and set quic as default

jacderida committed:
feat: provide `--local` flag for `add`

If this flag is set, the `--local` flag will be added to the arguments in the service definition.
This has been added to facilitate running a local network using services, which is going to be used
in the backward compatibility test in the `safe_network` repo.

jacderida committed:
feat: `status` command enhancements

Output the status report in json: this will be useful during CI and backwards compatibility testing
to obtain the multiaddr for the initial service node.

Fail if any nodes are not running: useful for quickly asserting that all nodes in the network must
be running for the test to proceed.

jacderida committed:
chore: provide script for local network

The difference between this and the `run` command is that it sets up the nodes as services rather
than just running them as processes. It requires root to run.

The script will be used by the local testnet Github Action to start a network for the backwards
compatibility test. The network will then be upgraded.

actions-user committed:
chore(release): 0.1.17

jacderida committed:
feat: provide `faucet` command

This command is useful to quickly spin up a faucet for a service-based local network, and it will be
used in the backwards compatibility test.

jacderida committed:
chore: fixup after rebase

An earlier refactor to return a node rather than a multiaddr has made it unnecessary to have the
`build_multiaddr` function in the node manager.

actions-user committed:
chore(release): 0.1.18

RolandSherwin committed:
fix: add delay to make sure we drop the socket

RolandSherwin committed:
test: force skip validation
- enable this once validation passes succesfully

RolandSherwin committed:
chore(deps): update sn_protocol to the newer version

actions-user committed:
chore(release): 0.1.19

RolandSherwin committed:
chore(test): remove unused structs

RolandSherwin commented on .gitignore:
```diff
@@ -1,12 +1,16 @@
 # Generated by Cargo
 # will have compiled files and executables
+<<<<<<< HEAD
```
Is this from a conflict?

RolandSherwin committed:
fix(deps): use local dependency for node manager

RolandSherwin committed:
refactor(manager): rename node manager crate

RolandSherwin committed:
fix(manager): increase port unbinding time

jacderida committed:
chore: reference crates in workspace

Use references to crates in the workspace rather than version numbers.

jacderida committed:
chore: rename sn_node_manager crate

The crate is renamed back to `sn-node-manager` because we've already had many versions published
with that name, but the directory will still be named `sn_node_manager` to be consistent with the
rest of the directories in the `safe_network` workspace.

jacderida committed:
ci: build the node manager on network start

When launching a local network, for now, we need to build the node manager rather than download the
latest version. Once the node manager has been added to the `safe_network` release process, we can
change this.

The `sn-local-testnet-action` has been updated to assume that we're building the node manager from
`safe_network`, as opposed to its own repository.

RolandSherwin committed:
fix(ci): specify custom CARGO_TARGET_DIR for windows test bin

RolandSherwin committed:
fix(ci): stop cargo test from building after the build step

jacderida committed:
ci: use main repo for local testnet action
